### PR TITLE
Switch from IPython to Jupyter

### LINKED ISSUE #1378: upgrade ipython2 to jupyter
We are using a modified `notebook_runner.py` which worked with IPython 2.

The idea is to upgrade this to the current one so that it can work with IPython 3 a.k.a. Jupyter.

I don't know how difficult this is but I'll start tracking links here.

If we had this we could wrap rcloud.python's `rcloud.language.support` function in a generator which takes a language kernel name, and then generate language supports instead of only using it for python.

### PR DESCRIPTION
**Pull request by useless5771**
Jupyter integration is implemented using nbconvert API and its ExecutePreprocessor class which implementation was customized to support code completion, standard input prompts and reuse of kernel instances.

For details on configuration see README.MD file.

FIX #1378, #1334

### CODE REVIEW

**Comment by gordonwoodhull** [comment]
Wow, fantastic!

Per #2565, I think we should rename the package, since it should be able to support many languages soon. I don't know why anyone would want to use the old `rcloud.python` package, but I think we should leave it intact and create a new package.

`rcloud.jupyter`, I suppose, even though it's somewhat ambiguous.

Couple little comments on the source. Hope to test this with Dave's help over the next week.

File: rcloud.packages/rcloud.python/R/main.R [hunk 1]
```diff
@@ -1,59 +1,112 @@
-.start.python <- function(rcloud.session)
+#' 
+#' Configuration
+#' 
+#' Python version and library path are defined by the following properties:
+#'  * rcloud.python.path - path to Python installation that should be used
+#'  * python.extra.libs - additional Python lib directories (e.g. where Jupyter Python modules got installed)
+#'
```
Comment (gordonwoodhull) [inline rcloud.packages/rcloud.python/R/main.R]: Please mention these are `rcloud.conf` settings. (Not that we have many other kinds of properties.)

File: rcloud.packages/rcloud.python/R/main.R [hunk 2]
```diff
@@ -1,59 +1,112 @@
...
+  }
+  
+  inline_plots <- "%matplotlib inline"
+  paste0(retina, '\n', inline_plots)
+}
+
+#'
+#' Initializes Jupyter Adapter and stores reference to it in session with 'jupyter.adapter' key.
+#'
+.start.jupter.adapter <- function(rcloud.session)
```
Comment (gordonwoodhull) [inline rcloud.packages/rcloud.python/R/main.R]: please fix typo in function name

File: rcloud.packages/rcloud.python/README.md [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+
+### Python 3
+
+```{bash}
+sudo python3 -m pip3 install jupyter
+```
+
+### Configuring RCloud Jupyter integration
+
+RCloud exposes the following properties:
```
Comment (gordonwoodhull) [inline rcloud.packages/rcloud.python/README.md]: again, maybe "uses these settings from `rcloud.conf`"

**Comment by useless5771** [comment]
Thank you @gordonwoodhull  for you review, I addressed the typos and improved the documentation.

Regarding renaming the package/creating a new one. I thought of doing this as part of #2565 but of course since rcloud.python should remain I will do this now. 

Would you like the rcloud.jupyter package to be put into separate git repository as well?

Thanks

**Comment by gordonwoodhull** [comment]
I think it's fine to keep it here, but if you prefer to move it that's an option too. It's a fuzzy line "what is core RCloud?"

**Comment by useless5771** [comment]
I think it would make sense to put it to separate repository, but I would leave it for now where it is to avoid cross-repository commits as part of #2565. We can move it to separate repository after that.

**Comment by useless5771** [comment]
I am closing this PR, it will be superseded by a PR that adds new rcloud.jupyter package instead of editing existing one.

**Pull request closed without merging**